### PR TITLE
[Project] Add `tree` argument to `get_artifact`

### PR DESCRIPTION
**Pull request by TomerShor**
Add an optional argument to get an artifact with a specific `tree` (`producer_id`).

Additionally, improve the naming and docstring of `mlrun.utils.parse_artifact_uri` for better usability.

Related to https://jira.iguazeng.com/browse/ML-4974

### CODE REVIEW

File: mlrun/projects/project.py [hunk 1]
```diff
@@ -3148,7 +3148,7 @@ def deploy_function(
             mock=mock,
         )
 
-    def get_artifact(self, key, tag=None, iter=None):
+    def get_artifact(self, key, tag=None, iter=None, tree=None):
```
Comment (alonmr) [inline mlrun/projects/project.py:3151]: Add the tree to the docstring

File: mlrun/utils/helpers.py [hunk 2]
```diff
@@ -660,11 +660,18 @@ def dict_to_json(struct):
...
 
 def parse_artifact_uri(uri, default_project=""):
-    uri_pattern = r"^((?P<project>.*)/)?(?P<key>.*?)(\#(?P<iteration>.*?))?(:(?P<tag>.*?))?(@(?P<uid>.*))?$"
+    """parse artifact uri into project, key, tag, iter, tree
+    uri format: [<project>/]<key>[#<iter>][:<tag>][@<tree>]
+
+    :param uri:            uri to parse
+    :param default_project: default project name if not in uri
+    :returns: project, key, iter, tag, tree
+    """
```
Comment (alonmr) [inline mlrun/utils/helpers.py:669]:
```suggestion
    """
    Parse artifact URI into project, key, tag, iter, tree
    URI format: [<project>/]<key>[#<iter>][:<tag>][@<tree>]

    :param uri:            uri to parse
    :param default_project: default project name if not in URI
    :returns: A tuple of:
              [0] = ...
              [1] = ...
    """
```

File: mlrun/utils/helpers.py [hunk 3]
```diff
@@ -660,11 +660,18 @@ def dict_to_json(struct):
...
+    uri_pattern = r"^((?P<project>.*)/)?(?P<key>.*?)(\#(?P<iteration>.*?))?(:(?P<tag>.*?))?(@(?P<tree>.*))?$"
```
Comment (alonmr) [inline mlrun/utils/helpers.py:670]: Not sure I follow why uid is no longer part of the artifact uri

**Review by alonmr — COMMENTED** [review]
A few comments. Also, I don't see `tree` in the get artifact endpoint params. Looks like it will be ignored.

File: mlrun/utils/helpers.py (hunk 3 above)
Comment (TomerShor) [reply mlrun/utils/helpers.py:670]: The `tree` was called `uid` before the refactor, and this is what actually was in the URI.
Now, UID is auto-generated in the BE, and this value was never a part of the URI.

**Comment by TomerShor** [comment]
> Also, I don't see `tree` in the get artifact endpoint params. Looks like it will be ignored.

@alonmr see https://github.com/mlrun/mlrun/blob/development/server/api/api/endpoints/artifacts_v2.py#L193
Since `project.py` is a client side function, it will only use the v2 endpoints.

**Pull request merged**